content/blog/2019-09-25-floats-static-arrays.md

```diff
@@ -0,0 +1,67 @@
+There are three operations related to manipulating arrays in Bril: initialization, assignment, and indexing.  Arrays are initialized with the `new` instruction, which takes a type and stores a default initialization of that type to the given destination.  Array elements can be set with the `set` instruction, an effectful instruction which takes the array, an index, and a value in that order.  Finally, array elements can be read using the `index` instruction, which writes a value of the given array element.
+
+Array operations are concisely summarized in the following code snippet:
+```
+val: int = const 5;
+ind: int = const 1;
+
+arr: int[3] = new int[3];   // arr = [0, 0, 0]
+set arr ind val;            // arr = [0, 5, 0]
+res: int = index arr ind;   // res = 5
+```
+Arrays can be written to freely in any Bril block.  As a result, array values are inherently stateful and should be treated by users carefully.  Writing values of the wrong type, however, is expected to not be permitted.
+
+### Implementation
+
+Floating points and fixed-length arrays have been implemented as described, and can be reasoned about by the `bril2json` parser, the `Brili` interpreter, and the `bril2txt` translator.  A subset of TypeScript can be written to Bril-recognized JSON by the `ts2bril` compiler.  While these conversions work generally as described above, there are some features which bear special notice.
+
+The `Brili` interpreter supports integers, doubles, and floating-points all as different objects.  Integers are represented as Javascript [BigInt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) objects.  This choice of representation means that `Brili` does not correctly reason about integer overflow -- while this should be implemented in the future, I have not observed a need for it in the current Bril projects.
+
+Doubles can be represented simply as `number`, while floats are derived with the [Math.fround](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/fround) function.  Note that `fround` is only applied _after_ each floating-point operation; this is permitted due to the [IEEE-754 restrictions](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html) on floating point operation precision loss (or lack of).  Extending Bril to support other mathematical operations (such as `pow` or `sqrt`), may require that this implementation choice be updated.
+
+TypeScript represents every number as a `double` by default, thus operations in TypeScript are translated to Bril as floating-point operations.  The TypeScript `bigint` type can be used to specially generate integers in Bril.  Operations between TypeScript `bigint` values are not supported at the time of this writing; however, this should be changed shortly.  Floating points cannot be easily represented by TypeScript code; however, I will look into using the `fround` function as a mechanism for generating Bril `float` values.
```

Comment:
```suggestion
TypeScript represents every number as a `double` by default, and thus operations in TypeScript are translated to Bril as floating-point operations.  The TypeScript `bigint` type can be used to specially generate integers in Bril.  Operations between TypeScript `bigint` values are not supported at the time of this writing; however, this should be changed shortly.  Floating points cannot be easily represented by TypeScript code; however, I will look into using the `fround` function as a mechanism for generating Bril `float` values.
```

In the first sentence, those are independent clauses, so they need a conjunction (and a comma)—not just "thus." https://capra.cs.cornell.edu/styleguide/#however

Reply:
> Operations between TypeScript `bigint` values are not supported at the time of this writing; however, this should be changed shortly.

What do you mean by this? I think `x + y` works if `x` and `y` are both `bigint`s—that's how the interpreter's integer values work. Maybe you mean when `x` is a `bigint` and `y` is a `number`?

If so, I'm not entirely sure this is relevant because, as you noted above, the types can't be mixed in operations.

Reply:
> Floating points cannot be easily represented by TypeScript code; however, I will look into using the `fround` function as a mechanism for generating Bril `float` values.

I think the previous paragraph says that you've already done this?